Add SYCL compiler and runtime library design documentation.

bader opened the pull request:
Signed-off-by: Alexey Bader <alexey.bader@intel.com>

Ruyk commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,259 @@
+- the offload wrapper object file
+- all the host libraries needed by the application
+- the SYCL runtime library
+The result is so-called "fat binary image" containing the host code, code for
```
Missing extra line between list and paragraph

Ruyk commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,259 @@
+## SYCL Runtime architecture
+*TBD*
+
+Supported extensions Intel subgroups
```
Probably 

```
## Supported extensions

* Intel subgroups [cl_intel_subgroups](https://www.khronos.org/registry/OpenCL/extensions/intel/cl_intel_subgroups.html)
```

bader replied:
Probably it should a link to the SYCL extension.
I'll add it once it available.

bader commented:
@Ruyk, thanks for your comments!

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,262 @@
+compiler generates SPIR-V and OpenCL device JIT compiler produces native target
+binary.
+
+There are existing options for OpenMP* offload:
```
I have the feeling you mean that you get the inspiration from OpenMP as an existing example, but it is not clear enough.
This suggests more that we can also have OpenMP as the same time as SYCL. But perhaps it just happens to work like that too? It would be great to be able to use OpenMP and SYCL in the same program, for example for HPC applications. Of course not OpenMP inside a SYCL kernel, except on a host queue.

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,262 @@
+target triples. This basically tells the driver which device compilers must be
+invoked to compile the SYCL kernel code. By default, the JIT compilation
+approach is assumed and device code is compiled for a single target triple -
+spirv-\*-\*.
```
`spirv-*-*`

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,262 @@
+#### Device code formats
+Each target may support a number of code forms, each device compiler defines
+and understands mnemonics designating a particular code form, for example
+"visa:3.3" could designate virtual ISA version 3.3 for Intel GPU target (Gen
```
"`visa:3.3`"

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,262 @@
+mean invoking device compiler multiple times. Multiple invocations are not
+needed if these forms can be dumped at various compilation stages by the single
+device compilation, like SPIR-V->visa->ISA. But if e.g. gen9:visa3.2 and
+gen9:visa3.3 are needed at the same time, then some mechanism is needed. Should
```
`gen9:visa3.2` and `gen9:visa3.3`

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,262 @@
+(per-target) separately and finally produce a "fat binary" - a host executable
+with embedded linked images for each target specified at the command line.
+
+![Multi source compilation flow](Multi-source-compilation-flow.png)
```
Could you put the figures in a vector source form such as SVG instead of binary bitmap?

keryell replied:
This is the last file still in PNG.

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,262 @@
+integration header is used (included) by the SYCL runtime implementation so it
+must be ready before host compilation starts.*
+
+Now, after all the device compilations are completed resulting in a_T2.bin and
```
All the file names in the following should be in TT font, such as `a_T2.bin`

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,262 @@
+(maybe on different machines).
+
+Device linking starts with breaking the fat objects back into constituents with
+the unbundler tool (bundler invoked with -unbundle option). For each fat object
```
`unbundle`

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,262 @@
+wrapping all the device images and provide the runtime with access to that
+information. So when creating the host wrapper object the offload wrapper tool
+does the following:
+- creates a .sycl_offloading.descriptor symbol which is a structure containing
```
`sycl_offloading.descriptor`

bader commented:
I've added a section about compiling C++ to SPIR-V.

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,377 @@
+  - TBD: potentially the middle-end optimizer can run any LLVM IR
+  transformation with only one limitation: back-end compiler should be able to
+  handle transformed LLVM IR.
+  - Optionally: LLVM IR -> SPIR-V translator.
```
Probably we can use directly the Unicode character here: →

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,377 @@
+target triples. This basically tells the driver which device compilers must be
+invoked to compile the SYCL kernel code. By default, the JIT compilation
+approach is assumed and device code is compiled for a single target triple -
+`[spir,spir64]-\*-\*`.
```
Probably the backslashes are not required here.

bader replied:
They are required, otherwise markdown will not show * because it thinks we are trying to make show `-` in italic text.

keryell replied:
Ah I see...

keryell replied:
Actually by looking at https://github.com/bader/llvm/blob/design-doc/sycl/doc/SYCL_compiler_and_runtime_design.md the backslashes are rendered...

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,377 @@
+**TBD:** Having multiple code forms for the same target in the fat binary might
+mean invoking device compiler multiple times. Multiple invocations are not
+needed if these forms can be dumped at various compilation stages by the single
+device compilation, like SPIR-V->visa->ISA. But if e.g. `gen9:visa3.2` and
```
→ →

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,377 @@
+![High level component diagram for SYCL Compiler](Compiler-HLD.svg)
+
+SYCL compiler logically can be split into the host compiler and a number of
+device compilers - one per each supported target. Clang driver orchestrates the
```
I have the feeling that here and in the following paragraphs most of your `-` here are actually emdashes instead `—`, except for the itemize syntax at the beginning of the lines.

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,377 @@
+
+The compiler supports linking of device code obtained from different source
+files before generating the final SPIR-V to be fed to the back-end. The basic
+mechanism is to produce "fat objects" as a result of compilation - object files
```
Is this some itemize list that was broken by some unexpected paragraph reformating in the source file?

bader replied:
I don't think so. Could you clarify what formatting do you think would be better, please?

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,377 @@
+targets requested via the command line - `T2`. When doing this first
+compilation, the driver requests the device compiler to generate an
+"integration header" via a special option. Device compilation for other targets
+\- `T1` - don't need to generate the integration header, as it must be the same
```
I guess that if you replace these `-` by some `—` then you do not need to escape them anymore.

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,377 @@
+\- `T1` - don't need to generate the integration header, as it must be the same
+for all the targets.
+
+*Design note: Current design does not use the host compiler to produce the
```
You suggest that it is just an implementation choice but, according to your good argumentation, it seems more like a fundamental reason to do so that just some implementation choice...

bader replied:
Actually I might imagine alternative design which doesn't require the header and breaks host/device compiler order.
For instance, device compiler might emit the same information as host side static (or dynamic) library. In this case we can postpone binding of the device code with the SYCL runtime to the link (or run) time.

keryell replied:
Perhaps...

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,377 @@
+such LLVM IR must comply to some special requirements. Unfortunately there is
+no canonical form of special built-in types and operations in LLVM IR, moreover
+we can't re-use existing representation generated by OpenCL C front-end
+compiler.  For instance here is how OpGroupAsyncCopy operation looks in LLVM IR
```
`OpGroupAsyncCopy`

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,377 @@
+
+- Type: `i32` --> `OpTypeInt`
+- Operation: `load` --> `OpLoad`
+- Calls: `call` --> `OpFunctionCall`
```
→ → →

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,377 @@
+corresponding equivalents in LLVM IR. E.g.
+
+- Type: ??? --> `OpTypeEvent`
+- Operation: ??? --> `OpGroupAsyncCopy`
```
→ →

keryell commented on sycl/doc/SYCL_compiler_and_runtime_design.md:
```diff
@@ -0,0 +1,377 @@
+- the type must be pre-declared as a C++ class in `cl::__spirv` namespace
+- the type must not have actual definition in C++ program
+- the operation is expressed in C++ as `extern` function not throwing C++
+  exceptions
```
I have the feeling that in sycl/include/CL/__spirv/spirv_ops.hpp there are still a lot of `noexcept` missing...

bader replied:
I guess this comment doesn't imply changes in the documentation, but rather in the source code. Right?

keryell replied:
Actually I do not remember what I wanted to say so long ago...

keryell commented:
@bader just a small effort to do on this document to make it great... :-)

bader commented:
@keryell, I've applied most of the comments, rebased and squashed all the commits.
The only thing left undone is conversion of multi-source-compilation-flow picture to SVG format.
@kbobrovs, do you have converted image?

bader commented:
I've pushed a new version with removed backslashes and added links to the extensions documentation.
